011-admin-server.md

```diff
@@ -0,0 +1,139 @@
+# Strimzi Admin Server
+
+The Strimzi Admin repository was setup in December 2019 to hold an implementation of an Admin API. The repository has not had any contributions yet so this proposal sets out what the structure of an Admin Server might look like in order to get to the initial implementation of the server.
+
+## Motivation
+
+An API server will:
+* provide a consolidated backend API for additional interfaces like a browser based UI or a CLI.
+* be capable of supporting both REST and GraphQL interfaces.
+* allow for more sophisticated APIs to be built on top of the existing APIs like the kubernetes API providing value-add to existing content.
+
+## Proposal
+
+### The server
+The proposal is to make a modular server using the Vert.x-Web toolkit which has strong support for creating both REST and GraphQL interfaces.
+The server contains a single Vert.x verticle which listens for inbound traffic on 1 or more configurable ports.
+Allowing multiple configurable ports will allow the server to expose different functionality on each port.
+A common use will be to separate traffic using different authentication protocols so, one port can listen for SCRAM users, and a different port can be used for OAuth users on a single admin server.
+All the APIs that are supported on the server will then be available on both of the two ports. 
+The server module itself does not contain any API implementations but defines a service interface using the Java SPI and will load modules that implement that interface using the Java Service Loader.
+The implementation of the APIs that the server then exposes can be created independently and will be automatically included in the running server if they are available to the Service Loader.
+
+The reason for the modular approach is that it allows the strimzi project to create a base set of functionality that can easily be extended by a downstream project.
```

Comment:
While I don't have a problem with this in principal, I think we need to ensure that the URL structure of the API reflects the identity of the service and also its version. This is to prevent two different plugins (e.g. from different vendors) using the same name and also to provide API evolution between different versions of the same plugin (it's not clear to me whether a service could provide more than one API version).

Reply:
@tombentley is it sufficient that the mount point cannot be empty (the root) and must be unique or were there other restrictions you had in mind? The versioning is a little more challenging because the world has moved away from versioning being present on the URI. A common way to version is to use the media type so, e.g. you would have `application/vnd.strimzi.v1+json`, `application/vnd.strimzi.v2+json`, etc and the type without versioning, i.e. `application/json` would always refer to the latest version. Do we want a registration object which allows an added service to specify its mount point, version and anything else that would be useful, e.g. a vendor identifier, but not include the version in the URI? This would of course imply that you couldn't have more than 1 version of the API loaded which makes sense as a true REST API has the same limitation. Handling different version is then the responsibility of the implementor of the API.

Reply:
I don't have a strong preference for including the version in the URL and TBH I wasn't aware of a trend to use media types instead. That said, I think there is _some_ value, for people how know Kubernetes' REST API, in doing something similar with api groups at least. So if it were _me_, I would be envisaging 

* `GET /` would return some description (presumably openapi) of the available endpoints
* The common, strimzi provided plugins might live under `/strimzi.io/...` or even `/core.strimzi.io/...`, to allow for other Strimzi api groups in the future.
* Plugins provided by some vendor might live under `/example.com/...`